postgresql/postgresql_plugin.cc

```diff
@@ -0,0 +1,832 @@
+}
+
+struct PostgresqlImportProcedure: public Procedure {
+
+    PostgresqlImportConfig procedureConfig;
+
+    PostgresqlImportProcedure(MldbServer * server,
+                         const PolyConfig & config,
+                         std::function<bool (Json::Value)> onProgress)
+        : Procedure(server)
+    {
+        procedureConfig = config.params.convert<PostgresqlImportConfig>();
+    }
+
+    pg_conn* startConnection(const PostgresqlImportConfig & config_) const
+    {
+        return startPostgresqlConnection(config_.databaseName, config_.host, config_.port);
+    }
+
+    virtual Any getStatus() const override
+    {
+        Json::Value result;
+        result["ok"] = true;
+        return result;
+    }
+
+    virtual RunOutput run(const ProcedureRunConfig & run,
+                          const std::function<bool (const Json::Value &)> & onProgress) const override
+    {        
+        RunOutput result;
+
+        auto runProcConf = applyRunConfOverProcConf(procedureConfig, run);
+
+        // Connect to Postgresl database
+        auto conn = startConnection(runProcConf);
+
+        // Create the output
+        std::shared_ptr<Dataset> output =
+        createDataset(server, runProcConf.outputDataset, nullptr, true ); //overwrite
+
+        //query the rows
+        auto res = PQexec(conn, runProcConf.postgresqlQuery.c_str());
+        if (PQresultStatus(res) != PGRES_TUPLES_OK)
+        {
+            string errorMsg = PQresultErrorMessage(res);
+            PQclear(res);
+            PQfinish(conn);
+            throw HttpReturnException(400, "Could not query PostgreSQL database", errorMsg);
+        }
+
+        int nfields = PQnfields(res);
+        int ntuples = PQntuples(res);
+
+        std::vector<std::pair<RowName, std::vector<std::tuple<ColumnName, CellValue, Date> > > > rows;
+        for(int i = 0; i < ntuples; i++) {
+            std::vector<std::tuple<ColumnName, CellValue, Date> > cols;
+            for(int j = 0; j < nfields; j++) {
+                cols.emplace_back(ColumnName(PQfname(res, j)), getCellValueFromPostgres(res, i, j), Date::Date::notADate());
+            }
+            rows.emplace_back(Path(string("row_" + std::to_string(i))), std::move(cols));            
+        }
+
+        output->recordRows(rows);
+
+        // Save the dataset we created
+        output->commit();
+
+        result = output->getStatus();
+
+        PQclear(res);
+        PQfinish(conn);
+        return result;
+    }
+};
+
+/*****************************************************************************/
+/* POSTGRES QUERY FUNCTION                                                   */
+/*****************************************************************************/
+
+struct PostgresqlQueryFunctionConfig
+{
+    string databaseName;
+    int port;
+    string host;
+
+    string query;
+
+    PostgresqlQueryFunctionConfig() {
+        databaseName = "database";
+        port = 1234;
```

Comment:
Isn't the default PostgreSQL port 5432?  I would make it a constexpr instead of copying the default in every config.